client/system/detect_cloud/detect.go

```diff
@@ -0,0 +1,92 @@
+package detect_cloud
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+)
+
+/*
+	This packages is inspired by the work of the original author (https://github.com/perlogix), but it has been modified to fit the needs of the project.
+	Original project: https://github.com/perlogix/libdetectcloud
+*/
+
+var hc = &http.Client{Timeout: 300 * time.Millisecond}
+
+func Detect() string {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	detectVendorFuncs := []func() string{
+		detectAlibabaCloud,
+		detectAWS,
+		detectAzure,
+		detectDigitalOcean,
+		detectGCE,
+		detectOracle,
+		detectIBMCloud,
+		detectSoftlayer,
+		detectVultr,
+	}
+
+	detectSoftwareFuncs := []func() string{
+		detectOpenStack,
+		detectContainer,
+	}
+
+	vendorResults := make(chan string, len(detectVendorFuncs))
+	softwareResults := make(chan string, len(detectSoftwareFuncs))
+
+	var wg sync.WaitGroup
+	wg.Add(len(detectVendorFuncs) + len(detectSoftwareFuncs))
```

Comment:
I would prefer to put wg.Add(1) inside the loops, so if there are any changes to this function we don't need to remember to modify this line.